### ISSUE TITLE
Warnings are not consistently displayed

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit#2699]
Working on #2695 I realized I'd opened a rather large can of works around the display of warnings. I'm using this issue to try to summarize the problems and figure out a direction.

As background: warnings were added after multiple asserts and rely on the infrastructure created for multiple asserts. That's why they work well in conjunction with failures but not with other terminating events.

In this context, a terminating event for a test is one of the following:
 * An Error, i.e. an exception that is not one NUnit itself understands
 * A failure, caused by Assert.Fail or any assert that fails.
 * An Ignored event, caused by Assert.Ignore
 * An Inconclusive event, caused by Assert.Inconclusive or the failure of an Assume
 * A Pass event, caused by Assert.Pass

A Warning, of course, is not a terminating event.

The following table shows the problems I see when several warnings are followed by a terminating event __outside__ of a Multiple Assert block. As in...

```C#
Assert.Warn("Warning1");
Assert.Warn("Warning2");
Assert.Fail("Terminating Event!");
```

The action varies according to the nature of the terminating event.

Event | Displays | Test Result | Notes
------|----------|------------|-------
Error | Warnings plus error | Error     | Correct behavior
Fail    | Warnings plus failure | Failure  | Correct behavior
Ignore | Ignore message    | Ignored | Should display warnings
Inconclusive | Inconclusive message| Inconclusive | Should display warnings
Pass | Pass message | Pass | Should display warnings. Not clear whether this should override the warnings and Pass. Options: (1) Record as error (2)  record as warning (3) add an extra warning and pass.

If the terminating event occurs within a multiple assert block, the actions are slightly different...

Event | Displays | Test Result | Notes
------|----------|------------|-------
Error | Warnings plus error | Error     | Correct behavior
Fail    | Warnings plus failure (at end of block) | Failure  | Correct behavior
Ignore | Ignore message  | Error | Should display warnings. See discussion below.
Inconclusive | Inconclusive message| Error | Should display warnings. See discussion below.
Pass | Pass message | Error | Should display warnings. See discussion below.

Three of these events cause an error result. Assert.Pass, Assume failures, Assert.Inconclusive and Assert.Pass are not permitted inside a multiple assert block. This seems arbitrary in retrospect. I see two general approaches:
1. Let the events terminate execution immediately, within the block.
2. Save the results till the end of the block.

In both cases, we would need to have some logic to decide on the test result. Our existing end-of-block logic uses the "worst" result. Should we continue to do that or use the __last__ event to guide us. There's a lot of ambiguity here, which is why I disallowed those events within a multiple assert block in the first place. Nevertheless, I think it's worth discussing now.

**Comment by julealgon** [comment]
On the "Pass" event outside of `Multiple` blocks:
> Should display warnings. Not clear whether this should override the warnings and Pass. Options: (1) Record as error (2) record as warning (3) add an extra warning and pass.

I vote for considering this a `Pass`, and add the warnings. This is how mostly all other tools will behave (msbuild for instance): build succeeds (unquestionably), and warnings are added to it. You can consider warnings as blockers but that's controlled by an opt-in flag, `TreatWarningsAsErrors`, that is `false` by default. 

Maybe if `Assert.Pass` was parameterized like this (`bool treatWarningsAsErrors = false`), it would allow consumers to decide on what they want: do they want to fail on warnings or not? Parameterization could happen on a bigger configuration level as well (so that someone running the tests from outside could see warnings as failures), and that global setting would override test-specific settings on each `Assert.Pass` call.

> Assert.Pass, Assume failures, Assert.Inconclusive and Assert.Pass are not permitted inside a multiple assert block. **This seems arbitrary in retrospect.**

I agree. It sounds at least unintuitive. I see no reason not to allow `Pass`/`Inconclusive` inside a multiple block. Even more so considering how `Assert.Multiple` was implemented in NUnit: using a single lambda. With a single lambda, you can have very complex branching logic, and I could totally see that in one of these branches, people would want to halt the test with a `Pass` or `Inconclusive`.

If removing these restrictions would make for a simpler codebase, I'm all for them. 

Also on this:

>  I see two general approaches:
> 
> 1. Let the events terminate execution immediately, within the block.
> 2. Save the results till the end of the block.

My vote goes to 2. It seems more intuitive to me to halt the test once one of these assertions are encountered, no matter if it is inside an `Assert.Multiple` or not. I know this is a bit vague but it's just my gut feeling.

Having said that, getting different behavior out of these specific assertion methods inside a `Multiple` block _could_ be confusing as well and considered nonorthogonal.

Any of the two options is better than just not supporting it though honestly.

**Comment by CharliePoole** [comment]
@julealgon I agree with you that we should treat pass as pass. (option 1) Just wanted to lay out what appear to me to be the options.

Treating warnings as failures seems to be an orthogonal issue. (Loosely, we can say "treat warnings as errors" except that NUnit handles errors differently from failures.) That is, if we treat warnings as failures, then we essentially don't have warnings but failures. OTOH, if the user doesn't select the option, then we have to handle both warnings and (real) failures, so the design has to allow for it.

My thinking is that we should clearly define "terminating events" and also define which terminating events become non-terminating while in a multiple assert block. That should all be laid out in a table in the docs. Easiest to understand for users would be if only unhandled errors terminated the block. However, that introduces the problem of multiple, conflicting terminating events within the block and we would need some logic to decide which one wins. What if the user executed `Assert.Pass` followed at some distance or depth in the code by an `Assert.Fail`. If we exit on the first event, then the test passes and the later code is never executed. If we save it all up to the end, then we would fail, but the detail would include a 'Pass' event. I'm not completely sure which is best, but I tend to think that calling Assert.Pass followed by Assert.Fail is some kind of a logic error and should not be allowed to happen.

**Comment by CharliePoole** [comment]
@nunit/framework-team This is been around for a while (2018) without a decision as to how it will work. With version 4 coming up, wouldn't this be a good time to decide, in case any breaking changes are needed?
